### ISSUE TITLE
Save files with creation timestamps

### ISSUE THREAD

**Issue by Hurgotron** [Simon-Tesla/RaccoonyWebEx#116]
I usually download images with wget, which manages to save the file with the creation timestamp from the server. Raccoony creates the files with the current timestamp. I would love to see an option which makes it possible to set the creation timestamp from the server like wget does.

If that's not possible, downloading a list of URLs instead of the files themselves would do - such a list could be fed to wget. While that may not be useful for all sites because of login requirement, e.g. FA allows downloads from direct links without login.

I guess this is not terribly useful for everyone, but in my opinion a proper archive should leave the files untouched, including timestamp :)

**Comment by Simon-Tesla** [comment]
The problem is that typically the metadata for the download gets scraped from the website before we tell the browser to download the file, and the destination filename is one of the parameters that needs to be supplied to that download API, so there isn't an easy way for the extension to do this without downloading the file twice. 

In theory we could start trying to scrape the site for an upload date if it has one, but parsing dates can get a bit hairy.

**Comment by Eupeptic1** [comment]
I did a little Googling and found that this same request was made to another plugin (DownThemAll), and one of its contributors claims that the WebExtension API (built into Firefox/Chrome) doesn't have a way to do it: https://github.com/downthemall/downthemall/issues/16 .

I experimented with a couple of my own submissions on FurAffinity, and when I directly download the image with `wget`, FA does seem to set the `Last-Modified` header to something close the original time that the file hit FA.  One I checked shows the human-readable time on FA as xx:48, but the HTTP header is xx:49:10, so it's not *exactly* right.  Raccoony could probably *read* that header from the browser, but apparently there's no way for Raccoony to provide a timestamp to the "download" part of the WebExtension API.

A possible workaround for @Hurgotron would require them to do some coding outside of Raccoony.  Turn on metadata files in Raccoony and use Raccoony interactively as normal, saving images of interest.  Then, go through the metadata files and pull the `url` field out of the JSON, because that should always be the direct link to the image, document, audio file, video file, etc.  Feed those URLs to `wget` and let it download the files again.  Put some sleeps in that script so you don't trip the rate-throttling or Cloudflare at the art site!  Also, this is obviously not great if you have metered bandwidth.

You could then keep the files with the "site" filenames as downloaded directly by `wget`, or use those files as a reference to set the timestamps on the files downloaded by Raccoony:

`touch -r path/to/file-downloaded-by-wget.png ~/Downloads/raccoony/furaffnity/file-downloaded-by-raccoony.png`

(I wanted to suggest something extremely clever like `wget --server-response --start-pos 1024m https://d.furaffinity.net/[...]`, to get just the headers without downloading the whole file again, but in that case, FA/Cloudflare doesn't send the `Last-Modified` header.  Darn.)

**Comment by Simon-Tesla** [comment]
I wonder if they support the HTTP HEAD method.

**Comment by Hurgotron** [comment]
Thank you, this may not be the most elegant solution, but it works well enough for me :)
I set the download to the original filename {siteFilenameExt} and then use this bash script in the download directory:

```
for url in $(grep '"url"' *.txt | grep -o '"http.*",' | sed -e 's/^"//' -e 's/",$//') ; do
        file=$(echo "$url" | grep -oE '[0-9]+\..*$')
        touch -d "`wget --server-response --spider "$url" 2>&1 | grep Last-Modified | sed -e 's/^.*Last-Modified: //'`" "$file"
done
```
Given the limitations of the WebExtension API, this seems to be the easiest (bandwidth saving) solution immediately available. - Closing the issue, thanks again!

**Comment by Eupeptic1** [comment]
I wondered if that command I proposed would make critters run away screaming, but apparently @Hurgotron knows what to do with it. :D

The only other thing I would suggest would be to add a `sleep 1` (or even longer), right after the touch command in that script.  That should greatly reduce the chance that any rate-limiting (Cloudflare or equal) that the art site may be doing will kick in.

I'm not sure about HEAD requests, but I *do* know that FA's Cloudflare will kick in if you do "too many" GET requests in a short period of time.  Probably the easiest way to see what this looks like is to bookmark a couple of dozen submissions at FA in a new bookmark folder, and then tell your browser to "open all in tabs" on that folder.  Some of them will open as normal and some of them will give you an error message from FA/Cloudflare.

Since you're using bash, wget, etc, I'm *guessing* that you are on some flavor of Linux.  If you happen to be on Windows, *and* you are using Google Chrome, Edge-based-on-Chromium, Chromium, or another Chrome-derived browser, there's one corner case to be aware of.  Chrome disallows tildes `~` in filenames on Windows; Raccoony changes them to underscores before attempting to save the file.  (Issue #90 was about this.)  In that particular case, the filename part of the URL from the metadata file and the filename on your disk might not always match.

In other words, for Chrome on Windows, you might see something like this in the metadata file
`"url": https://d.furaffinity.net/art/eupeptic/1234/5678.eupeptic_~~nifty~~art~~.png`
but the filename on your local disk will actually be
`5678.eupeptic___nifty__art__.png`
.  `touch` will probably end up creating an empty file named `eupeptic_~~nifty~~art~~.png`, instead of changing the timestamp on the `5678.eupeptic___nifty__art__.png` file that actually has the image in it.
